// bill/node.js
'use strict';

exports.__esModule = true;
exports.findAll = findAll;
exports.eachChild = eachChild;
exports.createNode = createNode;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodashArrayFindIndex = require('lodash/array/findIndex');

var _lodashArrayFindIndex2 = _interopRequireDefault(_lodashArrayFindIndex);

var _compat = require('./compat');

var _utils = require('./utils');

var NODE_TYPE = typeof Symbol === 'function' && Symbol['for'] && Symbol['for']('bill.node') || 0xadc3;

exports.NODE_TYPE = NODE_TYPE;
var NODE_TYPES = {
  COMPOSITE: 1,
  DOM: 2,
  TEXT: 3
};

exports.NODE_TYPES = NODE_TYPES;
function indexOfNode(arr, instOrElement) {
  return _lodashArrayFindIndex2['default'](arr, function (node) {
    return node.privateInstance === instOrElement || node.element === instOrElement;
  });
}

var isStale = _compat.ifDef({
  '>=15.2': function _(privateInstance) {
    if (!privateInstance) return false;
    if (privateInstance._hostNode !== null) return false;

    return privateInstance._renderedChildren == null;
  },
  '>=15': function _(privateInstance) {
    if (!privateInstance) return false;
    if (privateInstance._nativeNode !== null) return false;

    return privateInstance._renderedComponent == null;
  },
  '*': function _(privateInstance) {
    return privateInstance && privateInstance._rootNodeID === null;
  }
});

function normalizeSubject(subject) {
  return subject && !subject.getPublicInstance ? _utils.getInstanceFromNode(subject) || _utils.InstanceMap.get(subject) || subject : subject;
}

function findAll(subject, test, includeSelf) {
  if (!subject) return [];

  var found = [];
  var node = createNode(subject);

  // ReactEmptyComponents (return null; render <noscript/>) have null has their element
  // so don't assume it's there.
  if (includeSelf && node.element != null && test(node)) found = found.concat(node);

  return node.children.reduce(function (found, child) {
    return found.concat(findAll(child, test, true));
  }, found);
}

function eachChild(subject, fn) {
  var inst = undefined,
      element = undefined,
      publicInst = undefined;

  if (!subject) return;

  subject = normalizeSubject(subject);

  if (_react2['default'].isValidElement(subject)) return _react2['default'].Children.forEach(subject.props.children, function (child) {
    return child != null && fn(child);
  });

  if (!_utils.isReactInstance(subject)) return;

  inst = subject;
  element = inst._currentElement;

  if (inst.getPublicInstance) publicInst = inst.getPublicInstance();

  var renderedChildren = _utils.getRenderedChildren(inst, publicInst);

  if (_utils.isDOMComponent(publicInst)) {
    var child = element && element.props.children;

    // in cases where there is a single child
    // renderedChildren will be null if that child is a non-element
    // renderable thing, like a string or number.
    if (renderedChildren != null) Object.keys(renderedChildren || {}).forEach(function (key) {
      return fn(renderedChildren[key]);
    });else if (child != null && _utils.isTextElement(child)) fn(child);
  } else if (_react2['default'].isValidElement(element) && typeof element.type === 'function' && renderedChildren != null) {
    fn(renderedChildren);
  }
}

function createNode(subject, lastWrapper) {
  var node = undefined,
      element = undefined,
      inst = undefined;

  if (subject != null && subject.$$typeof === NODE_TYPE) return subject;

  subject = normalizeSubject(subject);

  if (_utils.isReactInstance(subject)) element = subject._currentElement, inst = subject;else element = subject, inst = null;

  var children = [],
      lastElement = undefined,
      type = undefined;

  if (element != null && element !== false) {
    if (_utils.isCompositeElement(element)) type = NODE_TYPES.COMPOSITE;else if (_utils.isDomElement(element)) type = NODE_TYPES.DOM;else if (_utils.isTextElement(element)) type = NODE_TYPES.TEXT;
  }

  node = Object.defineProperties({}, {
    $$typeof: { value: NODE_TYPE },
    nodeType: { value: type, enumerable: true },
    parentNode: { value: lastWrapper, enumerable: true },
    privateInstance: {
      enumerable: true,
      get: function get() {
        if (inst && isStale(inst)) inst = null;
        return inst;
      }
    },
    element: {
      enumerable: true,
      get: function get() {
        if (node.privateInstance) return node.privateInstance._currentElement;

        return element;
      }
    },
    instance: {
      enumerable: true,
      get: function get() {
        var publicInst = undefined;
        var privInst = node.privateInstance;

        if (!privInst) return;

        if (privInst.getPublicInstance) {
          publicInst = privInst.getPublicInstance();
          if (publicInst === null) publicInst = privInst._instance;
        } else if (_utils.isTextElement(node.element)) publicInst = node.element;

        return publicInst;
      }
    },

    findAll: {
      enumerable: true,
      value: function value(test, includeSelf) {
        return findAll(node, test, includeSelf);
      }
    },
    prevSibling: {
      enumerable: true,
      get: function get() {
        var children = lastWrapper ? lastWrapper.children : [],
            idx = indexOfNode(children, inst || node.element) - 1;

        return idx < 0 ? null : children[idx];
      }
    },
    nextSibling: {
      enumerable: true,
      get: function get() {
        var children = lastWrapper ? lastWrapper.children : [],
            idx = indexOfNode(children, inst || node.element) + 1;

        return idx >= children.length ? null : children[idx];
      }
    },
    children: {
      enumerable: true,
      get: function get() {
        var elem = node.element;
        if (!lastElement || elem !== lastElement) {
          if (children.length) children.length = 0;

          lastElement = elem;

          eachChild(node.privateInstance || elem, function (child) {
            child = createNode(child, node);

            if (!isStale(child.privateInstance)) children.push(child);
          });
        }

        return children;
      }
    }
  });

  return node;
}
